tiktoken 기반 코드 토큰화 API 추가

agliotomato opened the pull request:
## 연관된 이슈
#38 

## 작업 내용
1.  라이브러리 추가 - OpenAI 토크나이저인 tiktoken 의존성 추가
2. 요청 DTO와 OpenAPI 스펙에 include_tokens 옵션을 정의해 토큰 배열을 필요할 때만 응답에 포함하도록 했고, 모델→인코딩
    매핑은 get_encoding_name_for_model로 자동 추론한 뒤 실패할 경우 DEFAULT_ENCODING으로 폴백
3. 토큰 길이 초과(413)와 인코딩 오류 등을 처리하기 위해 에러 매퍼와 커스텀 에러를 확장했으며, 백엔드 전체 라우팅/미들웨
    어에도 코드 토큰화 서비스 주입
4. tiktoken 의존성을 정식 추가
## 스크린 샷

agliotomato committed:
feat: tokenizer 상수 및 DTO 추가

agliotomato committed:
feat:tokenizer 서비스 및 controller 구현

agliotomato committed:
feat: 라우트 및 미들웨어에 tokenizer API 엔드포인트 연결

agliotomato committed:
docs: code/tokenize API 문서 작성

agliotomato committed:
feat: 토크나이저 에러 처리 로직 확장

agliotomato committed:
chore: tiktoken 라이브러리 추가

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+      properties:
+        model:
+          type: string
+          description: "사용된 모델 이름(요청값 또는 기본값)"
```
불필요한 정보는 response해줄 필요 없습니다.

leehyowon14 commented on back-end/src/middlewares/auth.middleware.js:
```diff
@@ -6,6 +6,7 @@ const _publicPaths = [
   "/api/v1/auth/login",
   "/api/v1/auth/register",
   "/api/v1/user/check-id",
+  "/api/v1/code/tokenize",
```
이거는 왜 추가해둔 것인가요??

agliotomato replied:
제거하겠습니다.

leehyowon14 commented on back-end/src/service/tokenizer.service.js:
```diff
@@ -0,0 +1,133 @@
+
+export const TokenizerService = {
+  async tokenize({ code, model, encoding, include_tokens = true }) {
+    if (!code || code.length === 0) {
```
line 98-104
dto에서 이미 검증되었기 때문에 굳이 다시 검증할 이유는 없습니다.

현재 입력값 검증의 책임은 dto에 있고,
서비스는 입력값 검증 책임이 없습니다.

leehyowon14 commented on back-end/src/service/tokenizer.service.js:
```diff
@@ -0,0 +1,133 @@
+    try {
+      tokenizer = getEncoder(resolved.encoding);
+    } catch {
+      throw ERROR.UNSUPPORTED_ENCODING();
```
encoding 변수에 따라 발생하는 에러면, encoding을 enum으로 처리하여 dto에서 검증 가능할 듯 합니다.

leehyowon14 commented on back-end/src/service/tokenizer.service.js:
```diff
@@ -0,0 +1,133 @@
+
+    const tokensPayload = include_tokens ? Array.from(tokens) : undefined;
+
+    return {
```
리턴도 dto를 사용합시다!

leehyowon14 commented on back-end/src/dto/code/tokenize.request.dto.js:
service에서 검증하는 것들 중 여기서 처리할 수 있는 것들은 여기서 처리합시다

leehyowon14 commented on back-end/src/service/tokenizer.service.js:
code.tokenizer.service.js로 변경할 필요가 있을 것 같습니다!

agliotomato replied:
혹시 말씀하신 것이 파일명을 `code.tokenizer.service.js`로 바꾸라는 의미인지,
아니면 경로를 `src/code/tokenizer.service.js`로 옮기라는 의미인지 확인 부탁드립니다!

현재 맥락상 후자로 이해하고 있습니다.

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+          type: string
+          description: "토큰화할 코드 문자열"
+          maxLength: 200000
+        model:
```
클라이언트측에서 모델 정보를 알고 있을 필요도 없고,
클라이언트측에서 모델을 결정할 책임도 없습니당

모델은 서버에서 상수로 관리하는게 맞을거 같아요
불피요한 필드인거 같습니다.

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+        model:
+          type: string
+          description: "선택 사항. 예: gpt-4o-mini"
+        encoding:
```
인코딩도 모델과 마찬가지입니다.
클라이언트에서 결정할 이유가 없습니다.

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+        encoding:
+          type: string
+          description: "선택 사항. 직접 인코딩(o200k_base 등)을 지정할 수 있습니다."
+        include_tokens:
```
응답에 토크나이즈된 결과를 포함할 필요는 없을 듯 합니다.

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+        model:
+          type: string
+          description: "사용된 모델 이름(요청값 또는 기본값)"
+        encoding:
```
마찬가지 입니다.

leehyowon14 commented on back-end/apiSpec.yml:
```diff
@@ -323,6 +353,50 @@ components:
+          type: array
+          items: { type: integer }
+          description: "토큰 ID 배열(include_tokens=true인 경우에만 포함)"
+        preview:
```
tokens와 preview 모두 삭제해도 될 것 같습니다.

client측에게 정보를 과하게 제공하는 것 또한 잠재적 보안 취약점에 포함됩니다.


token_count 외에는 크게 필요치 않은 정보같습니다.
fallback 사용은 오히려 에러로 발생시켜야 할 것 같습니다.

leehyowon14 reviewed (changes requested):
엔드포인트 token이 아니라 api/v1/code/에 POST로 처리해주십쇼!!

Commit:
docs(api): tokenize API 스펙 단순화

- 요청에서 model, encoding, include_tokens 필드 제거
- 응답에서 model, encoding, used_fallback, tokens, preview 제거
- 응답은 token_count만 포함하도록 변경

Commit:
refactor(dto): tokenize 요청 DTO 필드 제거

- model, encoding, include_tokens 필드 제거
- code 필드만 유지하여 서버에서 모델/인코딩 관리

Commit:
fix(dto): tokenize 요청 DTO 오류 처리 개선

- 모든 필드에 타입 검증 및 커스텀 에러 메시지 추가
- zod 기본 에러 대신 ERROR 매퍼 코드 사용
- code 필드 타입 검증 강화

Commit:
feat(dto): tokenize 응답 DTO 추가

- token_count만 포함하는 응답 DTO 생성
- 서비스 응답을 DTO로 변환하여 반환

Commit:
refactor(service): tokenizer 서비스 중복 검증 제거

- DTO에서 이미 검증된 값에 대한 중복 검증 제거 (line 98-104)
- 응답을 DTO로 변환하여 반환
- 로직 단순화 및 책임 분리

Commit:
refactor(service): tokenizer 서비스 파일명 변경

- tokenizer.service.js → code.tokenizer.service.js로 변경
- 모든 import 경로 업데이트

Commit:
refactor(controller): code controller에서 response DTO 사용

- 서비스 응답을 DTO로 변환하여 반환하도록 수정

Commit:
refactor(middleware): auth middleware public path 정리

- /api/v1/code/tokenize public path 확인 및 제거

Commit:
refactor(service): 사용하지 않는 import 제거

- DEFAULT_ENCODING import 제거

leehyowon14 commented:
제가 커밋한거 한번 리뷰해주시고 크게 문제되는거 없으면 머지하셔도 될 것 같습니다!
